File: app/controllers/Move.rb
class AbstractMove
  attr_accessor :player
  attr_accessor :type
  
  def color
    (player == 1)? :white :
    (player == 2)? :black : nil
  end
end

class Move < AbstractMove
  Peg = 1
  Swap = 2
  Resign = 3
  Draw = 4
  Forfeit = 5
  Lost = 6
  
  attr_accessor :x
  attr_accessor :y

  def initialize(x, y, player)
    @x = x
    @y = y
    @player = player
    @type = Peg
  end
  
  def text
    letters = "abcdefghijklmnopqrstuvwxyz"
    letters[@x..@x] + String(@y + 1)
  end
end

class SwapMove < AbstractMove
  def initialize(player)
    @player = player
    @type = Move::Swap
  end

  def text
    "swap"
  end
end

class ResignMove < AbstractMove
  def initialize(player)
    @player = player
    @type = Move::Resign
  end

  def text
    "resign"
  end
end

class DrawMove < AbstractMove
  def initialize(player)
    @player = player
    @type = Move::Draw
  end
  
  def text
    "draw"
  end
end

class ForfeitMove < AbstractMove
  def initialize(player)
    @player = player
    @type = Move::Forfeit
  end

  def text
    "forfeit"
  end
end

class LostMove < AbstractMove
  def initialize(player)
    @player = player
    @type = Move::Lost
  end

  def text
    "lost"
  end
end

File: express/lib/domain/Move.js
// Translated from Move.rb

const PEG = 1;
const SWAP = 2;
const RESIGN = 3;
const DRAW = 4;
const FORFEIT = 5;
const LOST = 6;

class AbstractMove {
  get color() {
    if (this.player === 1) return 'white';
    if (this.player === 2) return 'black';
    return null;
  }
}

class Move extends AbstractMove {
  constructor(x, y, player) {
    super();
    this.x = x;
    this.y = y;
    this.player = player;
    this.type = PEG;
  }

  get text() {
    const letters = 'abcdefghijklmnopqrstuvwxyz';
    return letters[this.x] + String(this.y + 1);
  }
}

Move.Peg = PEG;
Move.Swap = SWAP;
Move.Resign = RESIGN;
Move.Draw = DRAW;
Move.Forfeit = FORFEIT;
Move.Lost = LOST;

class SwapMove extends AbstractMove {
  constructor(player) {
    super();
    this.player = player;
    this.type = SWAP;
  }
  get text() { return 'swap'; }
}

class ResignMove extends AbstractMove {
  constructor(player) {
    super();
    this.player = player;
    this.type = RESIGN;
  }
  get text() { return 'resign'; }
}

class DrawMove extends AbstractMove {
  constructor(player) {
    super();
    this.player = player;
    this.type = DRAW;
  }
  get text() { return 'draw'; }
}

class ForfeitMove extends AbstractMove {
  constructor(player) {
    super();
    this.player = player;
    this.type = FORFEIT;
  }
  get text() { return 'forfeit'; }
}

class LostMove extends AbstractMove {
  constructor(player) {
    super();
    this.player = player;
    this.type = LOST;
  }
  get text() { return 'lost'; }
}

module.exports = { Move, SwapMove, ResignMove, DrawMove, ForfeitMove, LostMove };
